ccm_web/server/src/auth/auth.controller.ts

```diff
@@ -0,0 +1,21 @@
+import { Request, Response } from 'express'
+import { Controller, Get, Req, Res, UseGuards } from '@nestjs/common'
+import { ApiExcludeEndpoint } from '@nestjs/swagger'
+
+import { JwtAuthGuard } from '../auth/jwt-auth.guard'
+import { AuthService } from './auth.service'
+
+@UseGuards(JwtAuthGuard)
+@Controller('auth')
+export class AuthController {
+  constructor (private readonly authService: AuthService) {}
+
+  @ApiExcludeEndpoint()
+  @Get('csrfToken')
+  async setCSRFTokenCookie (
+    @Req() req: Request, @Res({ passthrough: true }) res: Response
+  ): Promise<void> {
+    // Cookie options deliberately include defaults of httpOnly false and signed false.
+    res.cookie('CSRF-Token', req.csrfToken(), this.authService.commonCookieOptions)
```

Comment:
We could also just send the token in the response here. I'm seeing people talk about doing it either way.

Reply:
What does it mean to send it in the response?  It's a value that the client must add to subsequent requests itself?  Setting a cookie is like the automatic way of specifying values that must be included in requests.  I think that's a good way to go.

Reply:
Basically, I need to get this token to the client somehow. I'm not sure it matters particularly whether it's a cookie or in the response, because ultimately it needs to be added by client JS code to the `CSRF-Token` header for requests. Why? That's how CSRF protection seems to work. If the token were just read automatically from a cookie by the backend (like currently we're doing with the JWT), then it would also be vulnerable to CSRF attacks.

Reply:
Sure, I understand it needs to get to the client.

I think what you were saying in [your earlier comment](https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/148#discussion_r679589493) is that the token can be sent in a cookie as you're doing now, or "in the response".  I'm not certain what the latter entails.  However, it sounds as though it would require the client to do something special with it when it needs to make a request.  Using a cookie avoids those special steps.

Also, I think some of the other TL apps use cookies for CSRF and it works well.

Reply:
So it's either do what I'm doing now, in the client, pull the CSRF token from the non-`httpOnly` cookie (as I'm doing now in `client/src/api.ts`. Or return the csrfToken in `response.body` as we would any other data from a `GET` request. This would require a bit more code on the client-side, though we wouldn't need the `js-cookie` library I'm adding. Basically we'd do something like what the author of [this article](https://medium.com/@ryanchenkie_40935/react-authentication-how-to-store-jwt-in-a-cookie-346519310e81) is doing towards the end.

I'm honestly not sure what the security implications are for either approach. This [SO answer](https://stackoverflow.com/a/59549909) seems to suggest doing it as I'm doing it now, via cookie readable by the client.

Reply:
I looked at the code again and I see that I made a faulty assumption.  I was thinking that since the token was being set in a cookie, then that cookie would automatically be included when the UI makes a request to the API.  I see now where it is being set in the `CSRF-Token` header for each API request.

Now I understand better that the UI code **_must do something_**  to add the token to the request, regardless of the way it gets the token. So, the answer to your question whether the token should be set in a cookie is a little less obvious to me now.  I guess being able to simply get the token using `Cookies.get('CSRF-Token')` is pretty convenient.  Putting the token in a cookie might make it a little easier for an attacker to find it, so putting it in the response might provide more "security through obscurity".  I'm not sure, but I suspect if the token were sent as a part of the response, it would be a little more work to get it.

So, in the article by Chenkie you referenced, is the part you referred to the "[Getting and Setting the CSRF Token](https://medium.com/@ryanchenkie_40935/react-authentication-how-to-store-jwt-in-a-cookie-346519310e81#:~:text=getting%20and%20setting%20the%20csrf%20token)" section?  In there, it looks like the code is using a `csrf-token` service or API entry to get the token.  Would CCM need to do that as well?  Is that service or API entry provided by one of the modules you've installed or would that need to be added to CCM's code itself?

Reply:
Cool, sounds like we're more on the same page now. Yes, that is the section I'm referencing, although there is code above that that's relevant.

As far as I understand, `csurf` makes the token needed by the client available on an `express.Request` object, so it's up to the application programmer to get that token to the client (either through a cookie or in the response body). Yes, to use the response body, we'd have to call that endpoint and probably pass the CSRF token around the React components, instead of treating it more like a global (which I'm doing now). I can do that, but my current approach is probably less lines of code. The author gets around some of this by using an `axios` default header, but we're using `fetch`...

> Putting the token in a cookie might make it a little easier for an attacker to find it, so putting it in the response might provide more "security through obscurity". I'm not sure, but I suspect if the token were sent as a part of the response, it would be a little more work to get it.

This kinda makes sense; I'd be interested to see references supporting your argument. I'm guessing a sophisticated enough attacker could find it either way.

Reply:
@lsloan, I think pulling a CSRF token -- subsequently submitted via a custom header -- from a cookie using client code seems to be a standard approach, used by Angular for instance (https://angular.io/guide/http#security-xsrf-protection). See also https://en.wikipedia.org/wiki/Cross-site_request_forgery#Cookie-to-header_token

Because of the greater FE simplicity, I'm leaning towards not changing this. Please let me know if you have strong feelings that this is wrong or a significant security concern.